ncard_app/templates/predefinedQueries/phone_book.html

```diff
@@ -0,0 +1,97 @@
+{% extends 'events/base.html' %}
+{% load crispy_forms_tags %}
+
+{% block title %}Phone Book{% endblock %}
+
+{% block content%}
+    <h1>Phone Book</h1>
+
+        {% csrf_token %}
+    <style>
+        table, th, td {
+            border: 1px solid black;
+        }
+    </style>
+<table id="query-form" >
+    {% csrf_token %}
+    <tr>
+        <th> Title </th>
+        <th> Full Name </th>
+        <th> Organization Name </th>
+        <th> Phone(office) </th>
+        <th> Phone(Mobile) </th>
+        <th> Email </th>
+    </tr>
+    {% for ele in data %}
+        <tr>
+            <td>
+                {{ ele.primary_contact.title }}
+            </td>
+            <td>
+                {% if ele.primary_contact.surname_first %} {{ ele.primary_contact.surname }}
+                    {{ ele.primary_contact.given_name }}
+                {% else %}  {{ ele.primary_contact.given_name }}  {{ ele.primary_contact.surname }}
+                {% endif %}
+            </td>
+            <td>
+                {{ ele.name }}
+            </td>
+            <td>
+                {{ ele.primary_contact.contact.phone_office }}
+            </td>
+            <td>
+                {{ ele.primary_contact.contact.phone_mobile }}
+            </td>
+            <td>
+                {{ ele.primary_contact.contact.email }}
+            </td>
+        </tr>
+    {% endfor %}
+</table>
+<button onclick="get_export()">export</button>
+
+{% endblock %}
+
+
+{% block custom_js %}
+    <script>
+function get_export() {
+        const table = document.getElementById("query-form");
+        const data = []
+
+        for (var i = 0, rows = table.rows.length; i < rows; i++) {
+            for (var j = 0, cells = table.rows[i].cells.length; j < cells; j++) {
+                if (!data[i]) {
+                    data[i] = new Array();
+                }
+                data[i][j] = table.rows[i].cells[j].innerHTML
+            }
+        }
+        console.log(data)
+        const trans = JSON.stringify(data)
```

Comment:
When the user navigates to the phonebook page and clicks "export", this happens:

* The phonebook page is downloaded from the server
* The contents of the page are converted to JSON and uploaded to the server
* The JSON is converted to CSV and downloaded from the server

This wastes bandwidth and server processing power. It would be better to directly convert the phonebook to CSV, in the user's browser.

Reply:
Alternatively, inside predefinedQuery.py, you could get the data directly from Django instead of from the POST request contents.